### ISSUE TITLE
Fatal Exception: java.lang.NoSuchMethodError

### ISSUE THREAD

**Issue by thuat26** [realm/realm-java#3775]
#### Expected Results

> Bug fix

#### Actual Results

> Fatal Exception: java.lang.NoSuchMethodError: no static method with name='trace' signature='(Ljava/lang/String;[Ljava/lang/Object;)V' in class Lio/realm/log/RealmLog;
       at java.lang.Runtime.nativeLoad(Runtime.java)
       at java.lang.Runtime.doLoad(Runtime.java:418)
       at java.lang.Runtime.load(Runtime.java:328)
       at java.lang.System.load(System.java:511)
       at com.getkeepsafe.relinker.SystemLibraryLoader.loadPath(SystemLibraryLoader.java:29)
       at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:198)
       at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
       at com.getkeepsafe.relinker.ReLinker.loadLibrary(ReLinker.java:70)
       at com.getkeepsafe.relinker.ReLinker.loadLibrary(ReLinker.java:51)
       at io.realm.internal.RealmCore.loadLibrary(RealmCore.java:59)
       at io.realm.Realm.init(Realm.java:189)
       at com.xyz.controller.DatabaseHelper.setupDefaultConfig(DatabaseHelper.java:45)
       at com.xyz.MainApplication.onCreate(MainApplication.java:64)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1007)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4499)
       at android.app.ActivityThread.access$1300(ActivityThread.java:144)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1355)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:150)
       at android.app.ActivityThread.main(ActivityThread.java:5162)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:525)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:744)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
       at dalvik.system.NativeStart.main(NativeStart.java)

#### Steps & Code to Reproduce

> I can't reproduce this bug. But Crashlytics reported many crashes. 

#### Code Sample

```java

Realm.init(context);
RealmConfiguration configuration = new RealmConfiguration.Builder()
   .encryptionKey(KEY)
   .schemaVersion(7)
   .migration(new Migration())
   .build();
Realm.setDefaultConfiguration(configuration);

```

#### Version of Realm and tooling
Realm version(s): 2.1.1

Realm sync feature enabled: no
Android Studio version: 2.2.2

**Comment by cmelchior** [comment]
Hi @thuat26 
What devices are you seeing this on?

**Comment by thuat26** [comment]
@cmelchior I'm seeing this on lots of devices: Assus, Samsung... You can see more detail at here: http://crashes.to/s/dc79054f252

**Comment by kneth** [comment]
@zaki50 Can it be a ProGuard issue? See https://developer.android.com/training/articles/perf-jni.html#faq_FindClass

**Comment by zaki50** [comment]
@kneth If it's a ProGuard issue, @thuat26 can reproduce it easily.
I'll investigate it.

**Comment by zaki50** [comment]
That error log says that static `trace(String, Object[])` method is not found in `io.realm.log.RealmLog`.
However Realm-java 2.1.1 does not have a code that requires such a method when loading a native library.

We had such a code in Realm-java 2.0.x.
```cpp
JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM* vm, void*)
{
    JNIEnv* env;
    if (vm->GetEnv((void **) &env, JNI_VERSION_1_6) != JNI_OK) {
        return JNI_ERR;
    }
    else {
        g_vm = vm;
        // Loading classes and constructors for later use - used by box typed fields and a few methods' return value
        java_lang_long        = GetClass(env, "java/lang/Long");
        java_lang_long_init   = env->GetMethodID(java_lang_long, "<init>", "(J)V");
        java_lang_float       = GetClass(env, "java/lang/Float");
        java_lang_float_init  = env->GetMethodID(java_lang_float, "<init>", "(F)V");
        java_lang_double      = GetClass(env, "java/lang/Double");
        java_lang_double_init = env->GetMethodID(java_lang_double, "<init>", "(D)V");
        realmlog_class        = GetClass(env, "io/realm/log/RealmLog");
        log_trace             = env->GetStaticMethodID(realmlog_class, "trace", "(Ljava/lang/String;[Ljava/lang/Object;)V");
        log_debug             = env->GetStaticMethodID(realmlog_class, "debug", "(Ljava/lang/String;[Ljava/lang/Object;)V");
        log_info              = env->GetStaticMethodID(realmlog_class, "info", "(Ljava/lang/String;[Ljava/lang/Object;)V");
        log_warn              = env->GetStaticMethodID(realmlog_class, "warn", "(Ljava/lang/String;[Ljava/lang/Object;)V");
        log_error             = env->GetStaticMethodID(realmlog_class, "error", "(Ljava/lang/String;[Ljava/lang/Object;)V");
        log_fatal             = env->GetStaticMethodID(realmlog_class, "fatal", "(Ljava/lang/String;[Ljava/lang/Object;)V");
    }

    return JNI_VERSION_1_6;
}
```

I think that old native library is loaded by Realm Java 2.1.2.

**Comment by cmelchior** [comment]
Reason could be that we are not versioning the native code: https://github.com/KeepSafe/ReLinker#versioning

**Comment by cmelchior** [comment]
Fixed by #3785
